crates/provider-wit-bindgen/src/lib.rs

```diff
@@ -41,187 +55,926 @@ const EXPORTS_MODULE_NAME: &str = "exports";
+            wit_parser::Type::U8 => "u8",
+            wit_parser::Type::U16 => "u16",
+            wit_parser::Type::U32 => "u32",
+            wit_parser::Type::U64 => "u64",
+            wit_parser::Type::S8 => "i8",
+            wit_parser::Type::S16 => "i16",
+            wit_parser::Type::S32 => "i32",
+            wit_parser::Type::S64 => "i64",
+            wit_parser::Type::Float32 => "f32",
+            wit_parser::Type::Float64 => "f64",
+            wit_parser::Type::Char => "char",
+            wit_parser::Type::String => "String",
+            wit_parser::Type::Id(tydef) => {
+                let type_def = &cfg.wit_bindgen_cfg.resolve.types[*tydef];
+                if let Some(name) = &type_def.name {
+                    name.as_str()
+                } else {
+                    panic!(
+                        r#"
+Un-named type detected on argument while parsing interface [{}].
+Please ensure that types for exported interface functions (normally a single record type representing a message to be sent on the lattice) are specified.
+"#,
+                        iface.name.clone().unwrap_or("<unknown>".into()),
+                    )
+                }
+            }
+        };
+
+        let fn_name = Ident::new(iface_fn_name.to_snake_case().as_str(), Span::call_site());
+
+        let lattice_method = LitStr::new(
+            format!("Message.{}", iface_fn_name.to_upper_camel_case()).as_str(),
+            Span::call_site(),
+        );
+
+        let arg_name_ident = Ident::new(arg_name, Span::call_site());
+        let arg_type_ident = Ident::new(rust_type, Span::call_site());
+
+        let contract_ident = LitStr::new(&cfg.contract, Span::call_site());
+
+        Ok(quote::quote!(
+            async fn #fn_name(
+                &self,
+                #arg_name_ident: #arg_type_ident,
+            ) -> Result<(), ::wasmcloud_provider_sdk::error::ProviderInvocationError> {
+                let connection = ::wasmcloud_provider_sdk::provider_main::get_connection();
+                let client = connection.get_rpc_client();
+                let response = client
+                    .send(
+                        ::wasmcloud_core::WasmCloudEntity {
+                            public_key: self.ld.provider_id.clone(),
+                            link_name: self.ld.link_name.clone(),
+                            contract_id: #contract_ident.to_string(),
+                        },
+                        ::wasmcloud_core::WasmCloudEntity {
+                            public_key: self.ld.actor_id.clone(),
+                            ..Default::default()
+                        },
+                        #lattice_method,
+                        ::wasmcloud_provider_sdk::serialize(&#arg_name_ident)?,
+                    )
+                    .await?;
+
+                if let Some(err) = response.error {
+                    Err(::wasmcloud_provider_sdk::error::ProviderInvocationError::Provider(err.to_string()))
```

Comment:
Should this be `InvocationError`?

Reply:
Youd' think so, based on the code and comments like this one:

```rust
/// Errors that can occur when sending or receiving an invocation, including the `dispatch` method
/// of the provider.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum InvocationError {
```

But two things made me decide against that:

- Inclusion of error cases that clearly are about the host (ex. `Validation`/`Network` variants)
- `MessageDispatch` returns a `Result<Vec<u8>, ProviderInvocationError>`

Reply:
This is correct for now. It will be just a plain invocation error once I come back here to unflatten the rpc and provider errors